---
title: "RCAS_BRAF_scRNAseq_code"
author: "Lulu Xing"
date: "2023-11-05"
output: 
  html_document:
    toc: true
    toc_float: 
      toc_collapsed: FALSE
    toc_depth: 4
    theme: lumen
editor_options: 
  chunk_output_type: console
chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document of my first BD_Phapsody_scRNA.VDJ.

# 1. Load libraries and data

## 1.1 Install and load libraries 

```{r}
suppressMessages(library(Seurat))
suppressMessages(library(dplyr))
suppressMessages(library(ggplot2))
suppressMessages(library(devtools))
suppressMessages(library(SeuratData))
suppressMessages(library(Scillus))
```

## 1.2 Load data and check the data structure

```{r}
#setwd("~/Documents/Bioinformatics/BD Rhapsody")

#Read RDS object
p222a <- readRDS("~/Documents/Bioinformatics/BD Rhapsody/v2-rerun-p222a-1NS6000data-mWT-VDJ_Seurat.rds")

# check the data
#str(p222a)
table(Idents(p222a))
colnames(p222a@meta.data)

```

## 1.3 QC and normalization

```{r fig.height=8,fig.width=12, eval=TRUE}
#Subset out undetermined / multiplet cells 
table(p222a@meta.data$Sample_Name)

Idents(p222a) <- p222a@meta.data$Sample_Name
p222a <- subset(p222a, idents = c("Undetermined", "Multiplet"), invert = T) 
table(p222a@meta.data$Sample_Name)

#Calculate Percent.MT 
p222a[["percent.mt"]] <- PercentageFeatureSet(p222a, pattern = "^Mt.") 

#Visualize QC metrics / set threhsolds 
VlnPlot(p222a, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), pt.size = 0, group.by = "orig.ident") 

plot1 <- FeatureScatter(p222a, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(p222a, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

#Subset high quality cells based on QC thresholds 
p222a <- subset(p222a, subset = nFeature_RNA > 200 & nFeature_RNA < 9500 & percent.mt < 2) 
DefaultAssay(p222a) <- "RNA" 
dim(p222a)

```

# 2. Top variable genes and dimensionality reduction

## 2.1 SCT normalized assay

```{r fig.height=8,fig.width=12,eval=FALSE}
#Normalize mRNA and AbSeq Separately 
DefaultAssay(p222a) <- "RNA" 
p222a_normalized <- SCTransform(p222a) 

#save(p222a_normalized, file="~/Documents/Bioinformatics/BD Rhapsody/p222a_normalized.RData")
```

```{r fig.height=8,fig.width=12}

load("~/Documents/Bioinformatics/BD Rhapsody/p222a_normalized.RData")
dim(p222a)
dim(p222a_normalized)

#Set Default Assay to SCTransformed RNA data 
DefaultAssay(p222a_normalized) <- "SCT" 

#Identify Variable Features 
p222a_normalized <- FindVariableFeatures(p222a_normalized) 

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(p222a_normalized), 10)
#[1] 31855  8826
#[1] 28863  8826

```


```{r fig.height=8,fig.width=12}
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(p222a_normalized)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2

```

## 2.2 Scale normalized assay

```{r fig.height=8,fig.width=12, eval=FALSE}
#Scaling the data
all.genes <- rownames(p222a_normalized)
p222a_scaled <- ScaleData(p222a_normalized, features = all.genes)

dim(p222a_scaled)

#Run PCA
p222a_scaled <- RunPCA(p222a_scaled, features = VariableFeatures(object = p222a_scaled))

VizDimLoadings(p222a_scaled, dims = 1:2, reduction = "pca")
pca <- DimPlot(p222a_scaled, reduction = "pca",label=T, repel=T)

#Generate Elbow Plot 
ElbowPlot(p222a_scaled, ndims = 50) 

#Nearest Neighbor Graph Construction 
p222a_scaled <- FindNeighbors(p222a_scaled, dims = 1:20) 
p222a_scaled <- FindClusters(p222a_scaled, resolution = 0.4)

p222a_scaled <- RunTSNE(p222a_scaled , dims = 1:20) 
p222a_scaled <- RunUMAP(p222a_scaled, dims = 1:20) 

# rename clusters
Idents(p222a_scaled) <-"seurat_clusters"
new.cluster.ids <- c("Inflam TAMs","Reg TAMs","Tumor","Microglia","CSCs","MDMs","T cells",
                     "IFN TAMs","LA TAMs","ECs","BMDCs","B cells","OLs","VLMCs")

names(new.cluster.ids) <-levels(p222a_scaled)
p222a_scaled <-RenameIdents(p222a_scaled,new.cluster.ids)

#saveRDS(p222a_scaled, file = "~/Documents/Bioinformatics/BD Rhapsody/p222a_scaled_labelled.rds")
#save(p222a_scaled, file = "~/Documents/Bioinformatics/BD Rhapsody/p222a_scaled_labelled.RData")

```

## 2.3 Clustering analysis
```{r fig.height=8,fig.width=12}
p222a_scaled <- readRDS(file = "~/Documents/Bioinformatics/BD Rhapsody/p222a_scaled_labelled.rds")

library(Scillus)
#pdf("cluster_specific_BP_pathways.pdf", height=20, width=30)
plot_all_cluster_go(p222a.markers_cluster, org="mouse", ont="BP")
#dev.off()


Idents(p222a_scaled) <-"Sample_Name"
p222a_scaled2 <-subset(p222a_scaled, idents=c("SampleTag05_mm","SampleTag09_mm","SampleTag10_mm","SampleTag12_mm"))

Idents(p222a_scaled2) <-"Sample_Name"
levels(Idents(p222a_scaled2))

Idents(p222a_scaled2) <-"seurat_clusters"
new.cluster.ids <- c("Inflam TAMs","Reg TAMs","Tumor","Microglia","CSCs","MDMs","T cells",
                     "IFN TAMs","LA TAMs","ECs","BMDCs","B cells","OLs","VLMCs")
names(new.cluster.ids) <-levels(p222a_scaled2)
p222a_scaled2 <-RenameIdents(p222a_scaled2,new.cluster.ids)

tsne <-DimPlot(p222a_scaled2, reduction = "tsne",label=T, repel=T)
umap <-DimPlot(p222a_scaled2, reduction = "umap", label=T, repel=T) 

```

# 3. Find markers
## 3.1 Search markers
the process is slow, so I run the chunk, save the results and load in next chunk during the knit stage

```{r eval=FALSE}
p222a.markers_cluster <- FindAllMarkers(p222a_scaled, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#write.csv(p222a.markers_cluster, file="~/Documents/Bioinformatics/BD Rhapsody/p222a_markers_cluster.csv")

#save(p222a.markers_cluster,file="p222a_markers.RData")

```

## 3.2 Present markers
```{r fig.height=8, fig.width=12 }
load("p222a_markers.RData")
p222a.markers_cluster %>%
  group_by(cluster) %>%
  slice_max(n = 2, order_by = avg_log2FC)

p222a.markers_cluster %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC) -> top10

p222a.markers_cluster %>%
  group_by(cluster) %>%
  top_n(n = 5, wt = avg_log2FC) -> top5

```

## 3.3 check the genes of interests, feature plots
```{r fig.height=12,fig.width=12}
Idents(p222a_scaled2) <-"seurat_clusters"
RidgePlot(p222a_scaled2,features=c("Lgals3"), sort="increasing")


pdf("Galectin3_plot1_clusters.pdf", height=2, width=10)
VlnPlot(p222a_scaled2, features=c("Lgals3"), split.by="Sample_Name", assay="SCT")
dev.off()

pdf("Cd274_plot1_clusters.pdf", height=2, width=10)
VlnPlot(p222a_scaled2, features=c("Cd274"), split.by="Sample_Name", assay="SCT")
dev.off()

```

# 4. Subclustering analysis of T cells
## 4.1 Visualization 
```{r fig.height=12,fig.width=12}
Idents(p222a_scaled) <-"seurat_clusters"
Ts<-subset(p222a_scaled,idents=c("6"))
Ts <- readRDS(file = "~/Documents/Bioinformatics/BD Rhapsody/T cells cluster/Tcells.rds")

pdf(file="~/Documents/Bioinformatics/BD Rhapsody/featureplot_Tcells_markers.pdf", width=20,height=10)
FeaturePlot (Ts, features=c("Cd3g","Cd4","Cd8a","Pdcd1","Ctla4","Gzmb","Icos","Il2ra","Il2rb","Trac"), 
             reduction="tsne", min.cutoff="q10",max.cutoff="q90", order=TRUE)
dev.off()

#######
Ts <- FindClusters(Ts, resolution = 0.4)

Ts <- RunTSNE(Ts , dims = 1:20) 
Ts <- RunUMAP(Ts, dims = 1:20) 

tsne <-DimPlot(Ts, reduction = "tsne",label=T, repel=T)
umap <-DimPlot(Ts, reduction = "umap", label=T, repel=T) 

umap+tsne

tsne2 <-DimPlot(Ts, reduction = "tsne",label=T, repel=T,split.by="Sample_Name",ncol=2)
umap2<-DimPlot(Ts, reduction = "umap", label=T, repel=T,split.by="Sample_Name",ncol=2)

```

## 4.2 Search markers
the process is slow, so I run the chunk, save the results and load in next chunk during the knit stage

```{r eval=FALSE}
Tcellmarker_cluster <- FindAllMarkers(Ts, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
#write.csv(Tcellmarker_cluster, file="~/Documents/Bioinformatics/BD Rhapsody/Tcell_clusters_markers.csv")
#save(Tcellmarker_cluster,file="Tcell_markers.RData")

```

## 4.3 Present markers
```{r fig.height=8, fig.width=12 }
load("Tcell_markers.RData")

Tcellmarker_cluster %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC) -> top10

Tcellmarker_cluster %>%
  group_by(cluster) %>%
  top_n(n = 5, wt = avg_log2FC) -> top5

```

## 4.4 CD8 and CD4 T cell analysis
```{r fig.height=14,fig.width=12}
Cd8<-subset(Ts,idents=c("0"))
Idents(Cd8) <- "Sample_Name"
Cd8_2 <-subset(Cd8, idents=c("SampleTag05_mm","SampleTag09_mm","SampleTag10_mm","SampleTag12_mm"))


Cd4<-subset(Ts,idents=c("1"))
Idents(Cd4) <- "Sample_Name"
Cd4_2 <-subset(Cd4, idents=c("SampleTag05_mm","SampleTag09_mm","SampleTag10_mm","SampleTag12_mm"))

```

# 5 Clustering of other clusters
## 5.1 Cluster objects
```{r fig.height=14,fig.width=12}
Idents(p222a_scaled) <-"seurat_clusters"

Tumor <-subset(p222a_scaled,idents=c("2"))
Myeloid <-subset(p222a_scaled,idents=c("0","1","3","5","7","8"))
DCs <-subset(p222a_scaled,idents=c("10"))

Idents(Tumor) <- "Sample_Name"
Idents(Myeloid) <- "Sample_Name"
Idents(DCs) <- "Sample_Name"

Tumor <- readRDS(file = "~/Documents/Bioinformatics/BD Rhapsody/Tumor cluster/Tumor.rds")
Tumor2 <-subset(Tumor, idents=c("SampleTag05_mm","SampleTag09_mm","SampleTag10_mm","SampleTag12_mm"))

Myeloid <- readRDS("~/Documents/Bioinformatics/BD Rhapsody/Myeloid/Myeloid.rds")
Myeloid <-subset(Myeloid, idents=c("SampleTag10_mm","SampleTag05_mm","SampleTag09_mm","SampleTag12_mm"))

DCs <- readRDS(file = "~/Documents/Bioinformatics/BD Rhapsody/DCs/DCs.rds")
DCs <-subset(DCs, idents=c("SampleTag10_mm","SampleTag05_mm","SampleTag09_mm","SampleTag12_mm"))

```

## 5.2 Heatmaps 

```{r fig.height=8,fig.width=12,warning=FALSE, eval=FALSE}
library(pheatmap)
library(ComplexHeatmap)
library(circlize)

Tcellactivation1 <- c("Cd4","Cd3e","Tcf7","Cd69","Itgae","Itga1","Cd7","Il2ra",
                     "Tbx21","Cd27","Cd28","Ifng","Cxcr6","Cd44","Gzmk","Gzmb","Prf1",
                     "Lck","Klrg1","Pdcd1","Tnfrsf4","Tnfrsf18","Tnfrsf9")
Tcellactivation2 <- c("Cd8a","Cd3e","Tcf7","Cd69","Itgae","Itga1","Cd7","Il2ra",
                     "Tbx21","Cd27","Cd28","Ifng","Cxcr6","Cd44","Gzmk","Gzmb","Prf1",
                     "Lck","Klrg1","Pdcd1","Tnfrsf4","Tnfrsf18","Tnfrsf9")

Tcellexhaustion <- c("Cd274","Lag3","Ctla4","Havcr2","Tigit","Btla","Eomes","Lgals3","Fas","Fasl","Socs3","Cd276","Tnf","Foxp3")

#IFNg_genes <- c("GBP6","FGL2","EIF2AK2","IL2RB","STAT4","IRF1","NMI","UBE2L6","DHX58","USP18",
                #"IFITM3","CASP1","PSMB9","NOD1","CIITA","SOD2","B2M","SRI","CXCL10","CXCL11",
                #"SECTM1","IL6","SOCS3","CD74","DDX60","IFIT3","TRIM25","TRIM14","STAT1",
                #"XAF1","MX1","OAS2","OAS3","SLC25A28","IFI35","IRF7","HLA-DQA1","RTP4","DDX58",
                #"APOL6","PARP14","ZNFX1","SERPING1","RSAD2","HELZ2","CMPK2","STAT2","IFIH1",
                #"TNFSF10","HERC6","C1R","HLA-DMA","C1S","HLA-A","HLA-B","IFI27","NFKBIA",
                #"GBP4","HLA-DRB1","PARP12","ISG15","BST2","EPSTI1","PIM1","LGALS3BP","SAMHD1",
                #"IFIT2")

IFNg_genes2 <- c("H2-Eb2","H2-Aa","H2-Ab1","H2-DMb","Ifngr1","Ifngr2","Ifng","H2-M2",
                 "H2-Q6","H2-M3","H2-Q1","H2-Q2","H2-Q6","H2-Q4","H2-K1","H2-T23")

library(stringr)
tmp<-tolower(IFNg_genes2)
IFNg_genes2 <- str_to_title(tmp)

###########
Interest_genes1 <- c("NES","PROM1","GFAP","CLU","AQP4","CSPG4","PDGFRA","CNP","MBP","PLP1","PLLP","CLDN11",
                     "ASPA","OLIG2","EGFR","HES6","ASCL1","DLL1")
tmp<-tolower(Interest_genes1)
Interest_mm_genes <- str_to_title(tmp)

#############
cytokines <- c("Ccrl2","Il6ra","Il6","Il1b","Ccl3","Ccl2","Ccl9","Cxcl2","Il1rn","Ccl6","Il17ra","Il22","Cxcl3","Il9", "Il17","Il23","Ifng","Tgfb", "Lif","Lifr","Mmp9")


##############
ave_expt_genes <- AverageExpression(Cd8_2, features=Tcellactivation2)
ave_expt_genes <- AverageExpression(Cd8_2, features=Tcellexhaustion)
ave_expt_genes <- AverageExpression(Cd8_2, features=IFNg_genes2)

ave_expt_genes <- AverageExpression(Cd4_2, features=Tcellactivation1)
ave_expt_genes <- AverageExpression(Cd4_2, features=cytokines)
ave_expt_genes <- AverageExpression(Cd4_2, features=IFNg_genes2)

ave_expt_genes <- AverageExpression(Tumor2, features=cytokines)
ave_expt_genes <- AverageExpression(Tumor2, features=Interest_mm_genes)
ave_expt_genes <- AverageExpression(Tumor2, features=IFNg_genes2)
ave_expt_genes <- AverageExpression(Myeloid, features=IFNg_genes2)
ave_expt_genes <- AverageExpression(DCs, features=IFNg_genes2)

mat<-t(scale(t(as.matrix(ave_expt_genes$SCT)))) 
head(mat)
mat<-na.omit(mat)

col_fun = colorRamp2(breaks=c(-2, 0, 2),colors=c("blue","white", "red"))


pdf("ComplexHeatmap_Cd8_Tcells_exhaustion_markers.pdf", height=4, width=3)
Heatmap(mat, name = "Expression",  
        cluster_columns = TRUE,
        show_column_dend = TRUE,
        cluster_column_slices = TRUE,
        column_title_gp = gpar(fontsize = 1.5),
        column_gap = unit(0.5, "mm"),
        cluster_rows = TRUE,
        show_row_dend = TRUE,
        col = col_fun,
        row_names_gp = gpar(fontsize = 6),
        column_title_rot = 45,
        show_column_names = TRUE,
        use_raster = FALSE,
        raster_quality = 20)
dev.off()

```




    
